### ISSUE TITLE
Orca using mysql storage doesn't seem to respect the "Use Prior Execution" option for expected artifacts.

### ISSUE THREAD

**Issue by sgarlick987** [spinnaker/spinnaker#3667]
### Issue Summary:
Orca using mysql storage doesn't seem to respect the "Use Prior Execution" option for expected artifacts. I've executed the same pipeline using the redis storage without issues.

### Cloud Provider(s):
kubernetes v2 provider

### Environment:
halyard managed deploy onto kubernetes running in aws. orca sql configured using orca-local.yml
Image: gcr.io/spinnaker-marketplace/orca:1.1.0-20181003100130

### Feature Area: 
orca
pipelines
sql
artifacts

### Description:
I was testing using the sql storage engine for orca and noticed pipelines that had been working fine before were failing on new docker pushes. The pipeline itself is a simple bake -> deploy manifest, using a docker/webhook triggers, and a container, helm chart, and properties file as expected artifacts. 

I have all three artifacts set to use prior executions as the docker trigger only supplies docker artifact and the webhook supplies only the chart and properties artifacts. With redis this works as expected where a trigger off a docker registry includes the prior chart/properties artifacts, but with the sql storage that same trigger fails to include any of prior artifacts for the chart/properties and fails the pipeline since the expected artifacts are missing. I can switch my test env back and forth between redis and sql and the same pipeline gets the different results described.

### Steps to Reproduce:
1. Create a new pipeline with a docker registry, and webhook trigger.
2. Add docker, helm chart, and properties file artifact (http/file in jfrog for me specifically)
3. Set use prior execution on all expected artifacts
4. Set default artifact for docker container (latest in my case)
5. Add bake manifest step referencing the helm chart and properties artifacts
6. Add deploy manifest step using the artifact produced by the previous stage, and binding the docker artifact to it
7. Trigger the webhook trigger with  chart and properties artifacts
8. The pipeline should deploy fine at this point
9.  Trigger the pipeline by pushing a new version of the docker container
10. Pipeline fails due to missing the expected helm chart and properties artifact.

With the redis backend step 10 works as expected picking up the prior chart/properties artifacts.

Let me know if I can provide any other specific info. Thanks.

**Comment by ezimanyi** [comment]
@sgarlick987 : Is it possible that the reason is that the artifact exists in the prior executions in Redis but not in SQL?  Unless you specifically migrate the data from Redis to SQL (some info [here](https://www.spinnaker.io/guides/operator/orca-redis-to-sql/)), enabling SQL will just start with an empty execution history so won't find any artifacts.

If you are following all the steps and populating SQL execution history, and it's still not finding the artifacts, then this is probably a bug/oversight in the new SQL backend that we should look into.  (I'm not sure why this would be different for SQL as it looks like it just stores the full pipeline JSON in a column, but I'm not overly familiar with it so it's entirely possible there's something we need to fix there.)

**Comment by sgarlick987** [comment]
This is actually a fresh install so I didn't migrate the redis execution data (we had 2 pipelines for some non critical apps). I thought that might be the case at first, so I went ahead and flushed redis and cleared out sql before doing the steps to reproduce above again for both after I saw the initial issue to verify.

One other thing I noticed with sql is that the failed executions seemed to always go to the top of the list, even if I had another successful run that came after them. I didn't look into it much and not sure if its related but I'll open another issue if need be.

Like i said this install is pretty fresh and I don't mind breaking it if you need me to do any additional testing.

**Comment by sgarlick987** [comment]
Ok I took a look at that field for both run on step 8 and step 10. 

For the webhook one in step 8 that is successful, I see the default docker in the `trigger.artifacts`, and all the http/files in the `trigger.artifacts` and `trigger.payload` fields. 

For the docker triggered run after that fails in step 10 `trigger.artifacts` is empty.

I'm also in the slack channel if you need the actual json or anything, doesn't seem to be really anything sensitive, but I can clear out some of the identifying parts.

**Comment by ezimanyi** [comment]
It sounds based on this description like the trigger artifacts might not be persisting to the saved execution history for SQL...I'm not sure why that would be (as noted above it looks like the JSON blob is treated the same in redis or in SQL) but I'm not overly familiar with SQL.

I guess the minimal reproducible test case here (which is probably just a subset of what you've done) is to just run a pipeline with artifacts in the trigger with SQL enabled and confirm those artifacts don't show up in the execution history when you click 'view source' on that pipeline.

Maybe @robzienert has an idea why this would be the case, as he wrote most/all of the SQL stuff.

**Comment by robzienert** [comment]
This is really interesting. I don't have any installation that has artifact support, but the whole JSON payload will be the same Redis vs SQL. I'll look into getting an install with artifacts so I can dig some more here.

As for the sorting issue, that should be fixed, I seem to recall that being an issue early on in the SQL implementation, but it certainly shouldn't be the case anymore. I also recently merged an execution order change that should fix a few other little missteps.

**Comment by sgarlick987** [comment]
Great to hear, thanks for all the work you've been putting into this. Let me know if you need me to test any builds or anything.

**Comment by sgarlick987** [comment]
Thanks for testing it out. I'm going to try to dig a little deeper over the holidays in to what is going on and possibly upgrade as well to see if there's any difference.

**Comment by sgarlick987** [comment]
I got a chance to upgrade today and looks like still having the issue with a clean sql. I decided to try migrating to sql from a working redis setup and got some strange results. 

I started with a clean redis only install, executed the steps described before (webhook triggered artifacts + default docker artifact, then docker trigger + previous webhook artifacts) fine and can repeat either fine always picking up the right previous artifact.

I then went ahead and followed the instructions at https://www.spinnaker.io/guides/operator/orca-redis-to-sql/ and setup dual execution. This where things get a little strange. The next webhook trigger properly picked up the previous docker artifact from the redis only execution. Then I triggered it using the docker trigger, and it didn't pick up the artifacts from the previous sql execution, but instead the last execution from when it was setup for redis only. You can keep triggering by both webhook and docker from here and it'll always get the previous artifact from the redis only execution. 

If I'm understanding correctly with dual execution its read/writing to sql, and if we fail to read any executions it falls back to looking in redis? That seems to make sense from what I'm seeing above. 

Also all the runs after dual execution is enabled have all the artifacts populated in the pipeline json, just with the wrong values as described. Would that mean not that it's not matching previous artifacts, but not finding the sql executions at all?

**Comment by ezimanyi** [comment]
I'm somewhat baffled here...I just set up a test pipeline that's just a webhook accepting an artifact.  I ran it a few times in Redis-only mode with different artifacts, then switched to using the dual execution repository.  (I didn't set up migration, so the Redis executions stayed in Redis and the SQL executions stayed in SQL.)

I then ran the pipeline via the webhook trigger again, and it was persisted to SQL.  Then I manually started it and it correctly picked the most recent (SQL) execution to get the artifact.  I'm honestly not sure what could be happening differently in your setup...maybe it's related to the Redis->SQL migration if you set that up, but I'm really not familiar with that.

**Comment by robzienert** [comment]
I'm struggling to see how the migration would cause an issue, since it gives priority to the SQL executions (in this configuration) if they're present. Maybe writing out the behavior will help it click for me or someone else...

The [PipelineMigrationAgent](https://github.com/spinnaker/orca/blob/master/orca-migration/src/main/kotlin/com/netflix/spinnaker/orca/pipeline/persistence/migration/PipelineMigrationAgent.kt) is persistence agnostic; it'll behave the same for any migration. It works by scanning the old backend (Redis) for any Executions in a `COMPLETED` state. It's assumed by the migrator that once an Execution is `COMPLETED`, it will not be changing again, so it's a simple CAS operation against the new backend (SQL) based on the Execution Type & ID. When migrating, it copies the entire Execution as-is. Looking at [SqlExecutionRepository.storeExecutionInternal](https://github.com/spinnaker/orca/blob/master/orca-sql/src/main/kotlin/com/netflix/spinnaker/orca/sql/pipeline/persistence/SqlExecutionRepository.kt#L463), we only will update the `last_updated` field in SQL on this save; this is a SQL-only field not currently used by any application code.

I don't think it has anything to do with the migrators themselves.

Two theories:

1. It looks like the [ArtifactResolver](https://github.com/spinnaker/orca/blob/master/orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/util/ArtifactResolver.java#L139) uses its own sorting on an Rx Observable. I've seen odd behavior with Rx's `sort` before, so I wonder if in some cases we're getting irregular results here which may contribute to the issue?

2. `ArtifactResolver` does a list operation of previous Executions and in the `DualExecutionRepository` we'll perform a list against both Redis & SQL, merging the Observables together. For `ExecutionRepository` methods that have an `Observable<Execution>` return type, we don't run a `distinct()` on it, leaving that logic to higher-level code. I wonder if `ArtifactResolver`'s call to `retrievePipelinesForPipelineConfigId` and subsequent `findFirst` is returning the incorrect Execution due to repeated Executions? I wouldn't think so, since the results are all sorted?

**Comment by sgarlick987** [comment]
Sorry for the late reply been catching up on things for the new year. I'm going to work on getting a dev env setup Monday and see if I can't figure out whats wrong with our specific config. Thanks for all the info, it should help me get started tracking things down.

**Comment by sgarlick987** [comment]
Ok good news. I managed to get my local dev setup and reproduce the bug on 1.10.1. Even better when I switched to master to get ready to debug, turns out the issue was already fixed. I also went ahead and tested 1.11.6 (latest halyard version) and it seems to be fixed there as well. 

I'm going to test upgrading my cluster running in k8s just to be extra sure, but I'll go ahead and close this out then.

Thanks for everyones time.
